lib/kb_Metrics/metricsdb_controller.py

```diff
@@ -476,238 +473,239 @@ def get_user_job_states(self, requesting_user, params, token):
+        # ws_narratives = self.metrics_dbi.list_ws_narratives()
+        ws_narrs = copy.deepcopy(self.ws_narratives)
 
     def get_jobdata_from_ws_exec_ujs(self, params, token):
-	"""
-	get_jobdata_from_ws_exec_ujs--The original implementation to get data for appcatalog
-	from querying execution_engine, catalog, workspace and userjobstate
-	----------------------
-	To get the job's 'status', 'complete'=true/false, etc., we can do joining as follows
-	--userjobstate.jobstate['_id']==exec_engine.exec_tasks['ujs_job_id']
-	To join exec_engine.exec_apps with exec_engine.exec_tasks:
-	--exec_apps['app_job_id']==exec_tasks['app_job_id']
-	To join exec_engine.exec_apps with ujs.jobstate:
-	--ObjectId(exec_apps.app_job_state['job_id'])==jobstate['_id']
-	"""
-
-	# 0. get the ws_narrative and client_groups data for lookups
-	if self.ws_narratives is None:
-	    self.ws_narratives = self.metrics_dbi.list_ws_narratives()
-	if self.client_groups is None:
-	    self.client_groups = self.get_client_groups_from_cat(token)
-
-	# 1. query dbs to get lists of tasks and jobs
+        """
+        get_jobdata_from_ws_exec_ujs--The original implementation to get data for appcatalog
+        from querying execution_engine, catalog, workspace and userjobstate
+        ----------------------
+        To get the job's 'status', 'complete'=true/false, etc., we can do joining as follows
+        --userjobstate.jobstate['_id']==exec_engine.exec_tasks['ujs_job_id']
+        To join exec_engine.exec_apps with exec_engine.exec_tasks:
+        --exec_apps['app_job_id']==exec_tasks['app_job_id']
+        To join exec_engine.exec_apps with ujs.jobstate:
+        --ObjectId(exec_apps.app_job_state['job_id'])==jobstate['_id']
+        """
+
+        # 0. get the ws_narrative and client_groups data for lookups
+        if self.ws_narratives is None:
+            self.ws_narratives = self.metrics_dbi.list_ws_narratives()
+        if self.client_groups is None:
+            self.client_groups = self.get_client_groups_from_cat(token)
+
+        # 1. query dbs to get lists of tasks and jobs
         exec_tasks = self.metrics_dbi.list_exec_tasks(params['minTime'], params['maxTime'])
 
         exec_apps = self.metrics_dbi.list_exec_apps(params['minTime'], params['maxTime'])
 
-	ujs_jobs = self.metrics_dbi.list_ujs_results(params['user_ids'], params['minTime'], params['maxTime'])
-	ujs_jobs = self.convert_isodate_to_millis(ujs_jobs, ['created', 'started', 'updated', 'estcompl'])
+        ujs_jobs = self.metrics_dbi.list_ujs_results(params['user_ids'], params['minTime'],
+                                                     params['maxTime'])
+        ujs_jobs = self.convert_isodate_to_millis(ujs_jobs, ['created', 'started',
+                                                             'updated', 'estcompl'])
         return {'job_states': self.join_app_task_ujs(exec_tasks, exec_apps, ujs_jobs)}
 
-
     def join_app_task_ujs(self, exec_tasks, exec_apps, ujs_jobs):
```

Comment:
@qzzhang Please review this func very carefully. There are too many nested if else.